[apex] Expose more information on the nodes

adangel opened the pull request:
Like variable names, type names, apex version

This is a first suggestions. Feedback is welcome.
I've seen, that we have a helper class: https://github.com/pmd/pmd/blob/master/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/security/Helper.java from which many parts could be moved into the AST Nodes.

After that, we should review the rules to use the information available on the nodes, rather than let them dig into jorje.

adangel commented:
Some notes here:
* I think, it is beneficial, if the rules only use PMD's AST API rather than using directly Jorje internals. This means, we should make `ApexNode.getNode()` package-private.
* On `ASTLiteralExpression.getLiteralType()`, we expose a Jorje-internal Enum. Using this, makes the rule again dependent on jorje. Since there not so many different types, we could just provide for each type a getter, like the one I added now: `isString()`.
* Many nodes provide a some kind of type (type of a variable, field, return type of a method, ...). The naming is not consistent yet, e.g. `ASTField.getTypeRef()`, `ASTVariableDeclaration.getTypeName()`, `ASTMethod.getReturnType()`, `ASTProperty.getTypeName()`.
* the class `apex.rule.security.Helper` provides many utility methods for generating a fully qualified name of variables, fields, methods. They use the format "<defining_type>:<name>". I'm wondering, whether this actually works here in PMD, if Class A references Class B, but we let parse the files separately. I guess, the parser then doesn't resolve this?

adangel commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/security/ApexOpenRedirectRule.java:
```diff
@@ -102,26 +99,12 @@ private void findSafeLiterals(AbstractApexNode<?> node) {
-                 * sergey.gorbaty: Apex Jorje parser is returning a null from
-                 * Field.getFieldInfo(), but the info is available from an inner
-                 * field. DO NOT attempt to optimize this block without checking
-                 * that Jorje parser actually fixed its bug.
```
I've added a unit test in `ASTFieldTest`.

adangel commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/security/ApexOpenRedirectRule.java:
```diff
@@ -102,26 +99,12 @@ private void findSafeLiterals(AbstractApexNode<?> node) {
+                ASTField field = (ASTField) node;
+                if ("String".equalsIgnoreCase(field.getTypeRef())) {
+                    StandardFieldInfo fieldInfo = (StandardFieldInfo) field.getNode().getFieldInfo();
+                    if (fieldInfo.getValue() != null) {
```
In my tests, the value of the fieldInfo was always `null`... so, not sure, if this code is even executed.

adangel replied:
The problem was the quotes: Apex uses single quotes for Strings, not double quotes... with single quotes, there is a value.

oowekyala replied:
Is there an option in Jorje to make the parser fail on syntax errors rather than just produce an inconsistent AST? This would help save time if we enable that eg in tests and the designer

adangel replied:
Good question...
We currently run the compiler with stage "post type resolve": https://github.com/pmd/pmd/blob/f6549d6c0bbc896d9a27838dcc6eac0c5b97ce89/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/CompilerService.java#L76

There is also a stage called "validate" and "additional validate". Maybe one of these is what we want? I'll try this tomorrow.

adangel replied:
See #1741

adangel commented:
I've changed now the following:
* ASTLiteralExpression now supports all types
* Unified the *.getType() methods
* Most security rules are now also "jorje free". There is still work to do for ApexCRUDViolationRule and the Helper class. But that's for later.
* Updated wiki for potential PMD 7.0.0 changes: https://github.com/pmd/pmd/wiki/PMD-7.0.0#language-specific-improvements

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: No java rules are changed!
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No java rules are changed!</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
